suprb/optimizer/rule/ns/ns.py

```diff
@@ -138,171 +120,67 @@ def _optimize(self, X: np.ndarray, y: np.ndarray, n_rules: int) -> list[Rule]:
-            parents_w_ns = sorted(parents_w_ns, key=lambda x: (x[1], x[2]), reverse=True)
-            population.extend([x[0] for x in children_w_ns][:self.mu])
-            population.extend([x[0] for x in parents_w_ns][:n_rules])
-            return population
-
-    def _filter_for_minimal_criteria(self, rules: list[Rule]) -> list[Rule]:
-        # calculate the 25th percentile value and if it's lower than MNCS_threshold_matched use it instead to filter
-        # a maximum of 25% of the population (to prevent empty populations)
-        maximum_threshold = min(np.percentile([np.count_nonzero(rule.match_) for rule in rules], 
-                                              self.MCNS_threshold_matched), 25)
-        rules = [rule for rule in rules if np.count_nonzero(rule.match_) >= maximum_threshold]
-        return rules
-
-    def _filter_for_progressive_minimal_criteria(self, rules: list[Rule]) -> list[Rule]:
-        threshold_fitness = np.median([rule.fitness_ for rule in rules])
-        rules = [rule for rule in rules if rule.fitness_ >= threshold_fitness]
-        return rules
-
-    def _get_local_score(self, rule: Rule, rules_w_distances: list[tuple[Rule, float]]) -> float:
-        count_worse = 0
-        for x, _ in rules_w_distances[:self.NSLC_threshold]:
-            if x.fitness_ < rule.fitness_:
-                count_worse += 1
-        local_score = count_worse / len(rules_w_distances[:self.NSLC_threshold])
-        return local_score
-
-    def _get_pareto_front(self, rules_with_novelty_score: list[tuple[Rule, float, int]]) -> list[tuple[Rule, float,
-                                                                                                       int]]:
-        rules_with_novelty_score = sorted(rules_with_novelty_score, key=lambda a: (a[1], a[0].fitness_), reverse=True)
-        p_front = [rules_with_novelty_score[0]]
-
-        for r in rules_with_novelty_score[1:]:
-            if r[0].fitness_ >= p_front[-1][0].fitness_:
-                p_front.append(r)
-
-        return p_front
+        return parents
 
-    def _validate_params(self, n_rules: int):
-        if self.ns_type not in ["NS", "NSLC", "MCNS"]:
-            raise ValueError(f"{self.ns_type} is no valid NS-Type.")
-        if self.archive not in ["novelty", "random", "none"]:
-            raise ValueError(f"{self.archive} is no valid Archive-Type.")
-        if self.novelty_fitness_combination not in ["novelty", "50/50", "75/25", "pmcns", "pareto"]:
-            raise ValueError(f"{self.novelty_fitness_combination} is no valid Fitness-Novelty Combination.")
-        if n_rules > (self.mu + self.lmbda):
-            raise ValueError(f"n_rules={n_rules} must be less or equal to mu+lambda={self.mu+self.lmbda}.")
+    def _crossover_and_mutate(self, X: np.ndarray, y: np.ndarray, parents: list[Rule]) -> list[Rule]:
+        children = []
+
+        parent_combinations = zip(*[iter(parents)] * 2)
```

Comment:
wouldn't this allow us to select the same parent twice?

Reply:
No

![image](https://user-images.githubusercontent.com/45732695/183624746-cdd91a7d-bb2e-4848-b1df-0beccf481456.png)